pymc_marketing/mmm/lift_test.py

```diff
@@ -0,0 +1,366 @@
+
+    Returns:
+        Dictionary of indices for the lift test results in the model.
+
+    """
+    columns = df_lift_test.columns.tolist()
+    return {
+        col: _lift_test_index(df_lift_test[col].to_numpy(), np.array(model.coords[col]))
+        for col in columns
+    }
+
+
+def empirical_lift_measurements(x_before, x_after, saturation_curve, pt=pt):
+    """Calculate the empirical lift measurements at two spends.
+
+    Args:
+        x_before: Array of x before the change.
+        x_after: Array of x after the change.
+        saturation_curve: Function that takes spend and returns saturation.
+        pt: PyTensor module.
+
+    Returns:
+        Array of empirical lift measurements.
+
+    """
+    return pt.diff(
+        saturation_curve(pt.stack([x_before, x_after])),
+        axis=0,
+    ).flatten()
+
+
+def required_dims_from_named_vars_to_dims(
+    named_vars_to_dims: dict[str, tuple[str, ...]],
+) -> list[str]:
+    """Get the required dimensions from a named_vars_to_dims dictionary.
+
+    Args:
+        named_vars_to_dims: Dictionary of variable names to dimensions.
+
+    Returns:
+        List of required dimensions.
+
+    """
+    required_dims = set()
+    for dims in named_vars_to_dims.values():
+        for dim in dims:
+            required_dims.add(dim)
+
+    return list(required_dims)
+
+
+def indices_from_lift_tests(
+    df_lift_test: pd.DataFrame,
+    model: pm.Model,
+    var_names: list[str],
+) -> Indices:
+    """Get the indices of the lift test results in the model.
+
+    Args:
+        df_lift_test: DataFrame with lift test results.
+        model: PyMC model with arbitrary number of coordinates.
+        alpha_name: Name of the alpha parameter in the model.
+        lam_name: Name of the lambda parameter in the model.
+
+    Returns:
+        Dictionary of indices for the lift test results in the model.
+
+    """
+
+    named_vars_to_dims = {
+        name: dims
+        for name, dims in model.named_vars_to_dims.items()
+        if name in var_names
+    }
+
+    required_dims = required_dims_from_named_vars_to_dims(named_vars_to_dims)
+
+    for col in required_dims:
+        if col not in df_lift_test.columns:
+            raise KeyError(f"The required coordinates are {required_dims}")
+
+    return lift_test_indices(df_lift_test[required_dims], model)
+
+
+def index_variable(
+    var_dims: tuple[str, ...],
+    var: pt.TensorVariable,
+    indices: Indices,
+) -> pt.TensorVariable:
+    """Index the TensorVariable based on the required lift test indices."""
+    idx = tuple([indices[dim] for dim in var_dims])
+    return var.__getitem__(idx)
+
+
+def add_lift_measurements_to_likelihood(
+    df_lift_test: pd.DataFrame,
+    variable_mapping,
+    saturation_function,
+    model: Optional[pm.Model] = None,
+    dist=pm.Gamma,
+    name: str = "lift_measurements",
+) -> None:
+    """Add lift measurements to the likelihood of the model.
+
+    Args:
```

Comment:
I think we are following the numpy docstring format right? https://numpydoc.readthedocs.io/en/latest/format.html